plugin, config: make license key entirely optional

dgw opened the pull request:
Tin. Built on #4, but takes it to a new extreme as discussed in https://github.com/sopel-irc/sopel-iplookup/issues/3#issuecomment-1989629080.

Motivated by some joker on AWS repeatedly exhausting our account's daily download quota, spamming me with email notifications from MM, plus breaking any new installs of Sopel 7.x and this standalone plugin for 8.x.

Since the daily quota appears to be around 32 downloads, the existing download solution with an inbuilt key was never going to scale particularly well anyway. We might as well prepare now.

dgw committed:
plugin, config: MaxMind license key is now entirely optional

If no MaxMind key is provided, the plugin will download from a mirror of
the .mmdb files on GitHub.

With a license key, the plugin downloads directly from MaxMind.

I've also added error handling that didn't exist before, and though it's
not perfect, it's at least better. (There is now at least one case where
the function to find the GeoIP files can actually return False, which is
supposed to indicate an error to the call site inside the command func
but it was not actually possible to trigger that condition before.)

Note that I added the error handling because when I wrote this, the
project's MaxMind account was rate-limited for the day after some joker
on AWS used our whole download quota in a few minutes. Guess why I chose
to implement this change tonight?

dgw committed:
meta: add .gitignore file

Standard ignore file from the sopel-irc cookiecutter template. It's a
good'un, and cuts out all the egg-info/pycache/build-system cruft.

dgw commented:
Note: Starting off in draft state because I can't actually end-to-end test the download-with-key code path while our quota's exhausted. If abuse of the published key continues and I still can't test after a day or two, I'll see about creating a separate personal account at MM just for this. (Hopefully not necessary.)

dgw committed:
plugin: slightly better logging during/after DB downloads

dgw commented:
I successfully tested the direct-download code path at 1 minute past midnight UTC. Noticed a small opportunity for logging improvements, so did that too.

dgw commented:
Rebased after #4 merged.

SnoopJ commented on README.rst:
```diff
@@ -18,6 +18,27 @@ The recommended way to install this plugin is to use ``pip``::
+
+``sopel-iplookup`` can be configured by invoking Sopel's interactive wizard:
+
+    $ sopel-plugins configure iplookup
+    Configure Sopel GeoIP Lookup Plugin
+    Please consult sopel-iplookup's README to learn about its settings.
+
+    Path to existing GeoIP db files (leave empty to auto download): 
+    MaxMind license key (optional): (hidden input)
```
I think that's meant to be a code block?

````suggestion
```
$ sopel-plugins configure iplookup
Configure Sopel GeoIP Lookup Plugin
Please consult sopel-iplookup's README to learn about its settings.

Path to existing GeoIP db files (leave empty to auto download): 
MaxMind license key (optional): (hidden input)
```
````

dgw replied:
This is rST, not Markdown.

I'm not against changing the readme to Markdown, though. It'd be simpler to test-render with `grip` or a VS Code extension. (Couldn't get the one I found for rST to work.)

SnoopJ replied:
Sigh, right, my classic mistake 🤦

It does render incorrectly, though, I guess it needs an extra colon in the preceding paragraph

![image](https://github.com/sopel-irc/sopel-iplookup/assets/6125322/fc5d49d3-7fb1-43ba-ab65-37c61af2c6f7)

dgw replied:
Oh, so it does. Without the double colon it's taken as a blockquote, I guess.

dgw replied:
Should be fixed in the last push.

SnoopJ commented on README.rst:
```diff
@@ -18,6 +18,27 @@ The recommended way to install this plugin is to use ``pip``::
+    Path to existing GeoIP db files (leave empty to auto download): 
+    MaxMind license key (optional): (hidden input)
+
+If your OS distribution has GeoIP database files installed already, you can
+provide a filesystem path to the folder where they are stored. The plugin will
+auto-download the database files if it cannot find them locally.
+
+By default, GeoIP database downloads will use an automated mirror on GitHub. You
+can optionally provide your own MaxMind license key and the plugin will download
+directly from them—useful in case the mirror breaks, or if you would simply like
+to get database files from the source.
```
Hmm, the wording here bugs me a little bit. I think I'd be happier if the information here was in the reverse order. I.e. tell the user that the plugin has a sane default for GeoIP data, _then_ mention the alternate ways to supply a particular database.

dgw replied:
Really, at this point I'd be surprised if I managed to write something _without_ second-guessing the information order. Seems to be a tradition for me writing docs that you later review 😂

dgw replied:
This is going to come with a rewrite, since just swapping the paragraphs didn't flow very well.

SnoopJ commented on sopel_iplookup/plugin.py:
```diff
@@ -87,30 +95,58 @@ def _find_geoip_db(bot: SopelWrapper):
-                params=web.urlencode(dict(common_params, **{'edition_id': 'GeoLite2-%s' % edition})),
+    editions = ['ASN', 'City']
+
+    if config.ip.maxmind_license_key:
+        # Direct downloads via MaxMind account
+        base_url = 'https://download.maxmind.com/app/geoip_download'
+        common_params = {
+            'license_key': config.ip.maxmind_license_key,
+            'suffix': 'tar.gz',
+        }
+
+        for edition in editions:
+            geolite_urls.append(
+                '{base}?{params}'.format(
+                    base=base_url,
+                    params=web.urlencode(
+                        dict(
+                            common_params,
+                            **{'edition_id': 'GeoLite2-%s' % edition}
+                        )
+                    ),
+                )
             )
+    else:
+        # Indirect downloads with no key needed
+        base_url = (
+            'https://raw.githubusercontent.com'
+            '/P3TERX/GeoLite.mmdb/download/GeoLite2-{edition}.mmdb'
```
This looks like we should have a separate `_download_geoip()` helper (or two) to contain this new code

dgw replied:
Maybe. I chose not to be afraid of the extra indentation level, since much of the logic is shared. How about a `_get_geoip_urls()` helper?

SnoopJ replied:
I tend to drop leading `get` from helpers like that but I don't feel strongly about it. I don't mind the extra indentation level, but collapsing 24 lines to _"get the relevant download URL"_ feels like a win for legibility to me here.

dgw replied:
Functions and methods should always have verb-y names, in my book, but if it helps: I didn't use `get`. 😁

SnoopJ reviewed (changes requested):
Changes look generally good, I did have a few remarks.

dgw committed:
README: rewrite most of the "Configure" section

dgw committed:
plugin: move building download URLs into a separate generator function

Decompression function for .tar.gz files now has an early return in case
it is asked to decompress a file that is not an archive.

SnoopJ commented on README.rst:
```diff
@@ -30,14 +30,14 @@ Configure
+You can optionally provide your own MaxMind license key, and the plugin will
+download directly from them instead. Alternatively, if your OS distribution
+already has GeoIP database files kept up-to-date by the system package manager,
+you can provide the path where those files are stored.
```
Nice, I like this rewording. Thank you for indulging me 😁

SnoopJ reviewed (approved):
Nice one, let's do it :shipit:

dgw commented:
> Nice one, let's do it :shipit:

I'll shortly see about a PR w/automated release workflow triggered by `published` Releases on GH. Need anything else from me?

SnoopJ commented:
Nope, nothing from me

dgw commented:
As promised on IRC, I checked if this needed any squashing. I eliminated a fixup commit, but the rest of the history looked sane enough. Should be good to merge when you are back in decent-connectivity-land.

dgw commented:
@SnoopJ _Daily database download limit reached_ again. I'd like to delete the existing API key soon, but first we need a new release of this plugin available that doesn't need it. This patch does that.

I've been holding off because this plugin's nominally your baby, but I'm happy to merge and finish up a new release if you aren't able to spend time on it. 😸
